src/lib.rs

```diff
@@ -168,6 +170,11 @@ impl Decoder {
                     }
                 };
             }
+
+            if data.sz > 0 {
+                dav1d_data_unref(&mut data);
```

Comment:
This is IMHO a bug in the `decode()` API: if an error happens in `get_picture()` then we won't return that error but pretend all is fine, but the input data that was not parsed yet is just gone and can't be recovered anymore.

Reply:
I think it would make more sense to redesign the API here a bit:
 - `decode()` should be removed as it doesn't allow to recover from errors in a useful way
 - `send_data()` should wrap the input data (see also https://github.com/rust-av/dav1d-rs/pull/56 )
 - `send_data()` returns the remaining input data together with the `EAGAIN` errors so that the caller can pass it again as needed

I'll look into how to do that best tomorrow.

Reply:
Mh that also doesn't work because the data needs to stay valid after EAGAIN until the decoder has consumed it.

The refcount handling in dav1d makes not much sense to me.

Reply:
I think the Rust wrapper around the decoder needs to remember if the last call of `send_data()` was an `EAGAIN` and based on that keep the data alive internally and unref later as needed.

I'll think this through a bit more and make an API/implementation proposal.

Reply:
Yeah `send_data()` is also broken like this: if it returns `EAGAIN` you have lost the data, have no idea how much of it was read already, and can't really call it again with what it wants you to call it with.

I think `decode()` needs to go away and `send_data()` needs to return a way to call it again safely with the remainder.

Reply:
I think the assumption was that you never have to deal with EAGAIN since we work with full packets.
What is the behavior expected on EAGAIN? I think we should link the documentation if it exists.

Reply:
You have to call again with the same `Dav1dData` and it will continue consuming it, after collecting all output pictures that are ready currently.

I think what would make most sense here is to expose the behaviour as close as possible to the C API but in a way that is safe in Rust. Then error recovery etc can be dealt with by the caller in the way the caller works, and especially that means keeping `send_data()` and `get_picture()` split instead of having this chimera that is `decode()` currently.

I'll come up with a proposal for solving all this some time this week, probably this evening :)

Reply:
Converting to draft and making this the PR for providing a more useful / correct API.